tfhe/src/high_level_api/integers/types/base.rs

```diff
@@ -111,38 +111,76 @@ where
 
     pub fn cast_from<P2>(other: GenericInteger<P2>) -> Self
     where
-        P: IntegerParameter<InnerCiphertext = RadixCiphertext>,
-        P2: IntegerParameter<InnerCiphertext = RadixCiphertext>,
+        P2: IntegerParameter,
         P::Id: Default,
     {
         other.cast_into()
     }
 
-    pub fn cast_into<P2>(mut self) -> GenericInteger<P2>
+    pub fn cast_into<P2>(self) -> GenericInteger<P2>
     where
-        P: IntegerParameter<InnerCiphertext = RadixCiphertext>,
-        P2: IntegerParameter<InnerCiphertext = RadixCiphertext>,
+        P2: IntegerParameter,
         P2::Id: Default,
     {
         crate::high_level_api::global_state::with_internal_keys(|keys| {
             let integer_key = keys.integer_key.pbs_key();
             let current_num_blocks = P::num_blocks();
             let target_num_blocks = P2::num_blocks();
 
-            if target_num_blocks > current_num_blocks {
-                let num_blocks_to_add = target_num_blocks - current_num_blocks;
-                integer_key.extend_radix_with_trivial_zero_blocks_msb_assign(
-                    &mut self.ciphertext,
-                    num_blocks_to_add,
-                );
+            let blocks = if P::InnerCiphertext::IS_SIGNED {
+                if target_num_blocks > current_num_blocks {
+                    let mut ct_as_signed_radix =
+                        SignedRadixCiphertext::from_blocks(self.ciphertext.into_blocks());
```

Comment:
not sure I understand the from blocks and the check on InnerCiphertext IS_SIGNED and the need for the from_blocks, is it a matter of trait not knowing the concrete type ?

Reply:
Yes we don't know the concrete type, and extend_radix_with_sign_msb only works for SignedRadixCiphertext

Reply:
ok